### PR TITLE
BUG: Respect 'usecols' parameter even when CSV rows are uneven

### LINKED ISSUE #12203: io.read_csv: ValueError due to long lines
I try to read a csv file were some lines are longer than the rest.

Pandas `C engine` throws an error with these lines. 
But I do not was to skip these as discussed in [ a comment on a similar issue](https://github.com/pydata/pandas/issues/2886#issuecomment-30425534).

I prefer to "cut" the "bad columns" off using usecols. But I get the following errors:

```
df = pd.read_csv(file_path, sep=',', skiprows=1, header=None, 
                       usecols=range(0,23), nrows=None, 
                         engine='python')
```
Throws:

`ValueError: Expected 10 fields in line 100, saw 20`

Why is the `ValueError` raised although I explicitly defined the columns to use?

References:
* #2886 
* [Dealing with bad lines](http://github.com/pydata/pandas/issues/2886) -- does not apply
* [Dealing with bad lines II](http://nipunbatra.github.io/2013/06/reading-unclean-data-csv-using-pandas/)
* [Handling Variable Number of Columns with Pandas - Python](http://stackoverflow.com/questions/15242746/handling-variable-number-of-columns-with-pandas-python)
* [pandas read_csv and filter columns with usecols](http://stackoverflow.com/questions/15017072/pandas-read-csv-and-filter-columns-with-usecols), especially: http://stackoverflow.com/a/27791362

### PR DESCRIPTION
**Pull request by gfyoung**
Closes #12203 by overriding the row alignment checks for both engines when the `usecols` parameter is passed into `read_csv`.

### CODE REVIEW

**Comment by jreback** [comment]
@gfyoung its not necessary for you to keep rebasing every time master is slightly updated.

**Comment by gfyoung** [comment]
@jreback : Oh whoops, sorry!  I have a small script that updates my branches (including `master` in my fork) whenever I see there are commits, and I forgot to specify to just update the `master` branch.

**Comment by jreback** [comment]
oh ok maybe just do it manually for a while

trying to get out another rc

**Comment by jreback** [comment]
the issue is that Travis runs and it can get very busy

**Comment by gfyoung** [comment]
No, I perfectly understand.  As a precaution, I'll add `[ci skip]` to all of my open PR's.

File: pandas/io/tests/test_parsers.py [hunk 1]
```diff
@@ -2664,6 +2664,40 @@ def test_empty_header_read(count):
...
+        csv = r"""a,b,c
+        0,1,2
+        3,4,5,6,7
+        8,9,10
+        """
+
+        # make sure that an error is still thrown
+        # when the 'usecols' parameter is not provided
+        name = self.__class__.__name__
+        msg = "Expected \d+ fields in line \d+, saw \d+"
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: this is too much gymnastics here. Why is this needed? e.g. diffs in the errors from python/c parser? if that IS the case. then I'd like to change the raise for this kind of error to a ``ValueError`` in the cparser.

Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: Well that was why I had initially created a `PythonParserError` to keep it consistent with what you would see with the C engine.  `ValueError` though makes sense though for consistency purposes.  Will change.

Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: Actually, looking at where the `CParserError` is raised, it's buried in the `tokenizer.c` code, which is supposed to be for parsing errors in general.  Suggestions on how to externalize this particular `CParserError`?

Comment (jreback) [inline pandas/io/tests/test_parsers.py]: no its in ``parser.pyx`` (but it inherits from ``ValueError``). I would actually change that, but that might cause some issues (you can check that soln though)

Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: Oh, right, sorry!  Do you mean just getting rid of `CParserError` altogether and replacing with just `ValueError`?

Comment (jreback) [inline pandas/io/tests/test_parsers.py]: you wont' be able to remove it (well *shouldn't* as its in the public API). But it inherites from ``Exception``. if you can make it inherit from ``ValueError`` then I think we'd take that. See what happens. lmk.

Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: Fair enough.  That seems like a good compromise because `ValueError` inherits from `Exception` anyways.  Will give it a shot.

**Comment by gfyoung** [comment]
@jreback : Travis has no problems making `CParserError` inherit from `ValueError`, so I think this is also good to merge if there is nothing else.

**Comment by gfyoung** [comment]
@jreback : Travis has no problems making `CParserError` inherit from `ValueError`, so I think this is also good to merge if there is nothing else.

**Comment by jreback** [comment]
need s note in the API changes section wrt error class

**Comment by gfyoung** [comment]
Added note in the API changes.

**Comment by gfyoung** [comment]
Added note in the API changes.  Should be good to merge if there is nothing else.

File: doc/source/whatsnew/v0.18.1.txt [hunk 2]
```diff
@@ -50,6 +50,12 @@ API changes
 
 
 
+- ``CParserError`` is now a ``ValueError`` instead of just an ``Exception``
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: add the issue number for the PR here

File: doc/source/whatsnew/v0.18.1.txt [hunk 3]
```diff
@@ -90,6 +96,7 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
+- 'usecols' parameter is now respected even when the lines of a CSV file are not even (:issue:`12203`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: always say in ``pd.read_csv``

File: pandas/io/tests/test_parsers.py [hunk 4]
```diff
@@ -2664,6 +2664,39 @@ def test_empty_header_read(count):
...
+        # See gh-12203
+        csv = r"""a,b,c
+        0,1,2
+        3,4,5,6,7
+        8,9,10
+        """
+
+        # make sure that an error is still thrown
+        # when the 'usecols' parameter is not provided
+        name = self.__class__.__name__
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: extraneous?

Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: Oh, right.  Good catch.

File: doc/source/whatsnew/v0.18.1.txt (hunk 3 above)
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.1.txt]: Done.

File: doc/source/whatsnew/v0.18.1.txt (hunk 2 above)
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.1.txt]: Done.

**Comment by gfyoung** [comment]
@jreback : Made all of the requested changes.  Should be ready to merge if there is nothing else.

**Comment by gfyoung** [comment]
@jreback : Made all of the requested changes.  Should be ready to merge if there is nothing else.

**Comment by jreback** [comment]
thanks @gfyoung
